node/network/approval-distribution/src/lib.rs

```diff
@@ -1622,30 +1901,108 @@ impl State {
+					if !local && new_required_routing != RequiredRouting::GridXY {
 						metrics.on_aggression_l2();
-						*required_routing = RequiredRouting::GridXY;
+						new_required_routing = RequiredRouting::GridXY;
 					}
 				}
+				new_required_routing
 			},
+			&self.peer_views,
 		)
 		.await;
 	}
+
+	// Filter out invalid candidate index and certificate core bitfields.
+	// For each invalid assignment we also punish the peer.
+	async fn sanitize_v1_assignments(
+		&mut self,
+		peer_id: PeerId,
+		sender: &mut impl overseer::ApprovalDistributionSenderTrait,
+		assignments: Vec<(IndirectAssignmentCert, CandidateIndex)>,
+	) -> Vec<(IndirectAssignmentCertV2, CandidateBitfield)> {
+		let mut sanitized_assignments = Vec::new();
+		for (cert, candidate_index) in assignments.into_iter() {
+			let cert_bitfield_bits = match cert.cert.kind {
+				AssignmentCertKind::RelayVRFDelay { core_index } => core_index.0 as usize + 1,
+				// We don't want to run the VRF yet, but the output is always bounded by `n_cores`.
+				// We assume `candidate_bitfield` length for the core bitfield and we just check against
+				// `MAX_BITFIELD_SIZE` later.
+				AssignmentCertKind::RelayVRFModulo { .. } => candidate_index as usize + 1,
+			};
+
+			// Ensure bitfields length under hard limit.
+			if cert_bitfield_bits > MAX_BITFIELD_SIZE ||
+				cert_bitfield_bits != candidate_index as usize + 1
+			{
+				// Punish the peer for the invalid message.
+				modify_reputation(&mut self.reputation, sender, peer_id, COST_OVERSIZED_BITFIELD)
+					.await;
+			} else {
+				sanitized_assignments.push((cert.into(), candidate_index.into()))
+			}
+		}
+
+		sanitized_assignments
+	}
+
+	// Filter out oversized candidate and certificate core bitfields.
+	// For each invalid assignment we also punish the peer.
+	async fn sanitize_v2_assignments(
+		&mut self,
+		peer_id: PeerId,
+		sender: &mut impl overseer::ApprovalDistributionSenderTrait,
+		assignments: Vec<(IndirectAssignmentCertV2, CandidateBitfield)>,
+	) -> Vec<(IndirectAssignmentCertV2, CandidateBitfield)> {
+		let mut sanitized_assignments = Vec::new();
+		for (cert, candidate_bitfield) in assignments.into_iter() {
+			let cert_bitfield_bits = match &cert.cert.kind {
+				AssignmentCertKindV2::RelayVRFDelay { core_index } => core_index.0 as usize + 1,
+				// We don't want to run the VRF yet, but the output is always bounded by `n_cores`.
+				// We assume `candidate_bitfield` length for the core bitfield and we just check against
+				// `MAX_BITFIELD_SIZE` later.
+				AssignmentCertKindV2::RelayVRFModulo { .. } => candidate_bitfield.len(),
+				AssignmentCertKindV2::RelayVRFModuloCompact { core_bitfield } =>
+					core_bitfield.len(),
+			};
+
+			let candidate_bitfield_len = candidate_bitfield.len();
+			// Our bitfield has `Lsb0`.
+			let msb = candidate_bitfield_len - 1;
+
+			// Ensure bitfields length under hard limit.
+			if cert_bitfield_bits > MAX_BITFIELD_SIZE
+				|| cert_bitfield_bits != candidate_bitfield_len
```

Comment:
```suggestion
				|| cert_bitfield_bits != candidate_bitfield_len
```

CoreIndex is different from CandidateIndex, so this check will fail for RelayVRFDelay